src/update_validation.rs

```diff
@@ -4,78 +4,231 @@ use super::bootloader_env::bootloader_env::{
+
+        serde_json::to_writer_pretty(
+            OpenOptions::new()
+                .write(true)
+                .create(false)
+                .truncate(true)
+                .open(UPDATE_VALIDATION_COMPLETE_BARRIER_FILE)
+                .context("authenticated: write of {UPDATE_VALIDATION_COMPLETE_BARRIER_FILE}")?,
+            &self,
+        )
+        .context(
+            "authenticated: serializing of UpdateValidation to {UPDATE_VALIDATION_COMPLETE_BARRIER_FILE}",
+        )?;
+
+        // for now start validation blocking twin::init - maybe we want an successful twin::init as part of validation at some point?
+        self.check().await?;
+
+        Ok(())
+    }
+
+    async fn validate(&mut self) -> Result<()> {
+        debug!("update validation started");
+        let now = std::time::Duration::from(nix::time::clock_gettime(
+            nix::time::ClockId::CLOCK_MONOTONIC,
+        )?)
+        .as_millis();
+        let timeout_secs = u64::try_from(
+            (self.validation_timeout_ms - (now - self.start_monotonic_time_ms)) / 1000u128,
+        )?;
+        systemd::wait_for_system_running(timeout_secs).await?;
+
+        /* ToDo: if it returns with an error, we may want to handle the state
+         * "degrated" and possibly ignore certain failed services via configuration
+         */
+        info!("system is running");
+
+        // remove iot-hub-device-service barrier file and start service as part of validation
+        debug!("starting deviceupdate-agent.service");
+        fs::remove_file(UPDATE_VALIDATION_FILE).context("remove UPDATE_VALIDATION_FILE")?;
+
+        let now = std::time::Duration::from(nix::time::clock_gettime(
+            nix::time::ClockId::CLOCK_MONOTONIC,
+        )?)
+        .as_millis();
+        let timeout_secs = u64::try_from(
+            (self.validation_timeout_ms - (now - self.start_monotonic_time_ms)) / 1000u128,
+        )?;
+
+        systemd::start_unit(timeout_secs, IOT_HUB_DEVICE_UPDATE_SERVICE).await?;
+        debug!("successfully started iot-hub-device-update");
+
+        info!("successfully validated update");
+        Ok(())
+    }
+
+    async fn finalize(&mut self) -> Result<()> {
+        let omnect_validate_update_part = bootloader_env("omnect_validate_update_part")?;
+        ensure!(
+            !omnect_validate_update_part.is_empty(),
+            "omnect_validate_update_part not set"
+        );
+        set_bootloader_env("omnect_os_bootpart", omnect_validate_update_part.as_str())?;
+        unset_bootloader_env("omnect_validate_update")?;
+        unset_bootloader_env("omnect_validate_update_part")?;
+
+        fs::remove_file(UPDATE_VALIDATION_COMPLETE_BARRIER_FILE)
+            .context("remove UPDATE_VALIDATION_COMPLETE_BARRIER_FILE")?;
+        // cancel update validation reboot timer
+        self.tx
+            .take()
+            .unwrap()
+            .send(())
+            .expect("could not cancel update validation reboot timer");
```

Comment:
why `expect`?

Reply:
couldn't use anyhow context here

Reply:
just log an error, since the timer seems to be gone in that case.

Reply:
done